File: scripts/cycleOverlay.lua
require "const"

local CountScale = 3.0
local GScale = 1.5

local function update_time_slice (self, time)

   local v1 = 10
   local v2 = 10
   local v3 = 10

   local hil = dmz.object.hil ()

   if hil then
      if self.digitstate == 1 then
         if self.active > 0 then

            local state = dmz.object.state (hil)

            if state and state:contains (const.EngineOn) then

               local vel = dmz.object.velocity (hil)

               if not vel then vel = dmz.vector.new () end

               local speed = vel:magnitude () * 3.6;

               v1 = dmz.math.get_digit (speed, 0)
               v2 = dmz.math.get_digit (speed, 1)
               v3 = dmz.math.get_digit (speed, 2)
            end
         end
      else
         local count = 0
         if self.digitstate == 2 then
            count = dmz.object.counter (hil, const.WinsHandle)
         elseif self.digitstate == 3 then
            count = dmz.object.counter (hil, const.KillsHandle)
         elseif self.digitstate == 4 then
            count = dmz.object.counter (hil, const.DeathsHandle)
         end

         if not count then count = 0 end

         v1 = dmz.math.get_digit (count, 0)
         v2 = dmz.math.get_digit (count, 1)
         v3 = dmz.math.get_digit (count, 2)
      end
   end

   if v3 == 0 then
      v3 = 10
      if v2 == 0 then v2 = 10 end
   end

   dmz.overlay.enable_single_switch_state (self.digits[1], v1);
   dmz.overlay.enable_single_switch_state (self.digits[2], v2);
   dmz.overlay.enable_single_switch_state (self.digits[3], v3);

   if self.currentCount then
      local scale = dmz.overlay.scale (self.currentCount);
      if scale then
         scale = scale - (time * CountScale)
         if scale < 0.01 then scale = 0.01 end
         dmz.overlay.scale (self.currentCount, scale)
      end
   end

   if self.goactive then
      local scale = dmz.overlay.scale (self.gameover)
      if scale < GScale then
         scale = scale + (GScale * time);
         if scale > GScale then scale = GScale end
         dmz.overlay.scale (self.gameover, scale)
         local rot = dmz.overlay.rotation (self.gameover)
         rot = rot - (dmz.math.TwoPi * time * 3)
         if rot < 0 then rot = dmz.math.TwoPi + rot end
         dmz.overlay.rotation (self.gameover, rot)
      else
         dmz.overlay.scale (self.gameover, GScale)
         dmz.overlay.rotation (self.gameover, 0)
         dmz.overlay.enable_single_switch_state (self.gswitch, 2, true)
      end
   end

   if self.slider then
      if time > 0.1 then cprint (time) time = 0.1 end
      if self.dashstate then
         local x = dmz.overlay.position (self.slider)
         if x > 0 then x = x - (400 * time) end
         if x < 0 then x = 0 end
         dmz.overlay.position (self.slider, x, 0)
      else
         local x = dmz.overlay.position (self.slider)
         if x < 300 then x = x + (400 * time) end
         if x > 300 then x = 300 end
         dmz.overlay.position (self.slider, x, 0)
      end
   end
end

local function receive_input_event (self, event)

   if event.state then 
      if event.state.active then  self.active = self.active + 1
      else self.active = self.active - 1 end

      if self.active == 1 then
         self.timeSlice:start (self.handle)
      elseif self.active == 0 then
         self.timeSlice:stop (self.handle)
      end
   end

   if event.button then
      if event.button.which == 2 and event.button.value then
         self.dashstate = not self.dashstate
      elseif event.button.which == 3 and event.button.value then
         self.digitstate = self.digitstate + 1
         if self.digitstate > 4 then self.digitstate = 1 end
      end
   end
end

local function create_object (self, Object, Type)
   if Type:is_of_type (const.MCPType) then self.mcp = Object end
end

local function destroy_object (self, Object)

end

local function update_object_state (self, Object, Attribute, State, PreviousState)
   if Object == self.mcp then
      if not PreviousState then PreviousState = dmz.mask.new () end
      if State:contains (const.GameCountdown5)
            and not PreviousState:contains (const.GameCountdown5) then
         dmz.overlay.enable_single_switch_state (self.switch, 5)
         self.currentCount = self.countdown[5]
         dmz.overlay.scale (self.currentCount, CountScale)
         dmz.overlay.all_switch_state (self.waiting, false)
         dmz.overlay.all_switch_state (self.gswitch, false)
         self.goactive = false;
      elseif State:contains (const.GameCountdown4)
            and not PreviousState:contains (const.GameCountdown4) then
         dmz.overlay.enable_single_switch_state (self.switch, 4)
         self.currentCount = self.countdown[4]
         dmz.overlay.scale (self.currentCount, CountScale)
      elseif State:contains (const.GameCountdown3)
            and not PreviousState:contains (const.GameCountdown3) then
         dmz.overlay.enable_single_switch_state (self.switch, 3)
         self.currentCount = self.countdown[3]
         dmz.overlay.scale (self.currentCount, CountScale)
      elseif State:contains (const.GameCountdown2)
            and not PreviousState:contains (const.GameCountdown1) then
         dmz.overlay.enable_single_switch_state (self.switch, 2)
         self.currentCount = self.countdown[2]
         dmz.overlay.scale (self.currentCount, CountScale)
      elseif State:contains (const.GameCountdown1)
            and not PreviousState:contains (const.GameCountdown1) then
         dmz.overlay.enable_single_switch_state (self.switch, 1)
         self.currentCount = self.countdown[1]
         dmz.overlay.scale (self.currentCount, CountScale)
      elseif State:contains (const.GameActive)
            and not PreviousState:contains (const.GameActive) then
         dmz.overlay.all_switch_state (self.switch, false)
         self.currentCount = nil
      end

      if State:contains (const.GameWaiting)
            and PreviousState:contains (const.GameActive) then
         dmz.overlay.scale (self.gameover, 0.01)
         dmz.overlay.enable_single_switch_state (self.gswitch, 1, true)
         self.goactive = true;
      end
   end
end


local function start (self)
   self.handle = self.timeSlice:create (update_time_slice, self, self.name)

   self.inputObs:init_channels (
      self.config,
      dmz.input.Button + dmz.input.ChannelState,
      receive_input_event,
      self);

   if self.handle and self.active == 0 then self.timeSlice:stop (self.handle) end

   local callbacks = {
      create_object = create_object,
      destroy_object = destroy_object,
      update_object_state = update_object_state,
   }
   self.objObs:register (nil, callbacks, self)
end


local function stop (self)
   if self.handle and self.timeSlice then self.timeSlice:destroy (self.handle) end
   self.inputObs:release_all ()
end


function new (config, name)
   local self = {
      start_plugin = start,
      stop_plugin = stop,
      name = name,
      log = dmz.log.new ("lua." .. name),
      timeSlice = dmz.time_slice.new (),
      inputObs = dmz.input_observer.new (),
      objObs = dmz.object_observer.new (),
      active = 0,
      config = config,
      digits = {
         dmz.overlay.lookup_clone_sub_handle ("digit1", "switch"),
         dmz.overlay.lookup_clone_sub_handle ("digit2", "switch"),
         dmz.overlay.lookup_clone_sub_handle ("digit3", "switch"),
      },
      switch = dmz.overlay.lookup_handle ("countdown switch"),
      countdown = {
         dmz.overlay.lookup_handle ("one"),
         dmz.overlay.lookup_handle ("two"),
         dmz.overlay.lookup_handle ("three"),
         dmz.overlay.lookup_handle ("four"),
         dmz.overlay.lookup_handle ("five"),
      },
      slider = dmz.overlay.lookup_handle ("dashboard slider"),
      dashstate = true,
      digitstate = 1,
      waiting = dmz.overlay.lookup_handle ("waiting switch"),
      gameover = dmz.overlay.lookup_handle ("gameover transform"),
      gswitch = dmz.overlay.lookup_handle ("gameover switch"),
   }

   self.log:info ("Creating plugin: " .. name)
   
   return self
end

File: scripts/cycleOverlay.js
var dmz =
       { object: require("dmz/components/object")
       , time: require("dmz/runtime/time")
       , consts: require("const")
       , input: require("dmz/components/input")
       , portal: require("dmz/components/portal")
       , vector: require("dmz/types/vector")
       , defs: require("dmz/runtime/definitions")
       , matrix: require("dmz/types/matrix")
       , util: require("dmz/types/util")
       , overlay: require("dmz/components/overlay")
       , mask: require("dmz/types/mask")
       }

  , CountScale = 3.0
  , GScale = 1.5
  , SliderSpeed = 3.0
  , DigitState = 0
  , Active = 0
  , MCP = null
  , CurrentCount = null
  , GoActive = null

  , Active = 0
  , Digits =
       [ dmz.overlay.lookup("digit1").lookup("switch")
       , dmz.overlay.lookup("digit2").lookup("switch")
       , dmz.overlay.lookup("digit3").lookup("switch")
       ]
  , CDSwitch = dmz.overlay.lookup("countdown switch")
  , CountDown =
       [ dmz.overlay.lookup("one")
       , dmz.overlay.lookup("two")
       , dmz.overlay.lookup("three")
       , dmz.overlay.lookup("four")
       , dmz.overlay.lookup("five")
       ]
  , Slider = dmz.overlay.lookup("dashboard slider")
  , DashState = true
  , Waiting = dmz.overlay.lookup("waiting switch")
  , GameOver = dmz.overlay.lookup("gameover transform")
  , GSwitch = dmz.overlay.lookup("gameover switch")
  , SpeedSwitch = dmz.overlay.lookup("speed switch")
  , WinsSwitch = dmz.overlay.lookup("wins switch")
  , KillsSwitch = dmz.overlay.lookup("kills switch")
  , DeathsSwitch = dmz.overlay.lookup("deaths switch")

  , timeSlice
  , updateTimeSlice
  ;

updateTimeSlice = function (time) {

   var v0 = 10
     , v1 = 10
     , v2 = 10
     , hil = dmz.object.hil()
     , state
     , vel
     , speed
     , count
     , scale
     , rot
     , x
     ;

   if (hil) {

      if (DigitState === 0) {

         if (Active > 0) {

            state = dmz.object.state(hil);

            if (state && state.contains(dmz.consts.EngineOn)) {

               vel = dmz.object.velocity(hil);

               if (!vel) { vel = dmz.vector.create(); }

               speed = vel.magnitude() * 3.6;

               v0 = (speed % 10);
               v1 = (speed % 100) / 10;
               v2 = (speed % 1000) / 100;
            }
         }
      }
      else {

         count = 0;

         if (DigitState === 1) {

            count = dmz.object.counter(hil, dmz.consts.WinsHandle)
         }
         else if (DigitState === 2) {

            count = dmz.object.counter(hil, dmz.consts.KillsHandle)
         }
         else if (DigitState === 3) {

            count = dmz.object.counter(hil, dmz.consts.DeathsHandle)
         }

         if (!count) { count = 0 }

         v0 = (count % 10);
         v1 = Math.floor((count % 100) / 10);
         v2 = Math.floor((count % 1000) / 100);
      }
   }

   if (v2 === 0) {

      v2 = 10;

      if (v1 === 0) { v1 = 10; }
   }

   Digits[0].enableSingleSwitchState(v0);
   Digits[1].enableSingleSwitchState(v1);
   Digits[2].enableSingleSwitchState(v2);

   if (CurrentCount) {

      scale = CurrentCount.scale()[0];

      if (scale) {

         scale -= (time * CountScale);
         if (scale < 0.01) { scale = 0.01; }
         CurrentCount.scale(scale);
      }
   }

   if (GoActive) {

      scale = GameOver.scale()[0];

      if (scale < GScale) {

         scale += (GScale * time);
         if(scale > GScale) { scale = GScale; }
         GameOver.scale(scale);
         rot = GameOver.rotation();
         rot -= (Math.PI * 2 * time * 3);
         if (rot < 0) { rot += (Math.PI * 2); }
         GameOver.rotation(rot);
      }
      else {

         GameOver.scale(GScale);
         GameOver.rotation(0);
         GSwitch.enableSingleSwitchState(2, true);
      }
   }

   if (Slider) {

      if (DashState) {

         x = Slider.position()[0];
         if (x > 0) { x -= (400 * time * SliderSpeed); }
         if (x < 0) { x = 0; }
         Slider.position(x, 0);
      }
      else {

         x = Slider.position()[0];
         if (x < 300) { x += (400 * time * SliderSpeed); }
         if (x > 300) { x = 300; }
         Slider.position(x, 0);
      }
   }
};

dmz.input.channel.observe(self, function (channel, state) {

   if (state) {  Active += 1; }
   else { Active -= 1; }

   if (Active === 1) { timeSlice = dmz.time.setRepeatingTimer(self, updateTimeSlice); }
   else if (Active === 0) { dmz.time.cancleTimer(self, timeSlice); }
});

dmz.input.button.observe(self, function (channel, button) {

   if ((button.id === 2) && button.value) {

      DashState = !DashState;
   }
   else if ((button.id === 3) && button.value) {

      DigitState += 1;

      if (DigitState > 3) { DigitState = 0; }

      SpeedSwitch.enableSingleSwitchState(DigitState === 0 ? 0 : 1)
      WinsSwitch.enableSingleSwitchState(DigitState === 1 ? 0 : 1)
      KillsSwitch.enableSingleSwitchState(DigitState === 2 ? 0 : 1)
      DeathsSwitch.enableSingleSwitchState(DigitState === 3 ? 0 : 1)
   }
});

dmz.object.create.observe(self, function (obj, Type) {

   if (Type.isOfType(dmz.consts.MCPType)) { MCP = obj; }
});

dmz.object.state.observe(self, function (obj, Attribute, State, PreviousState) {

   if (obj === MCP) {

      if (!PreviousState) { PreviousState = dmz.mask.create(); }

      if (State.contains(dmz.consts.GameCountdown5)
            && !PreviousState.contains(dmz.consts.GameCountdown5)) {

         CDSwitch.enableSingleSwitchState(5);
         CurrentCount = CountDown[4];
         CurrentCount.scale(CountScale);
         Waiting.setSwitchStateAll(false);
         GSwitch.setSwitchStateAll(false);
         GoActive = false;
      }
      else if (State.contains(dmz.consts.GameCountdown4)
            && !PreviousState.contains(dmz.consts.GameCountdown4)) {

         CDSwitch.enableSingleSwitchState(4);
         CurrentCount = CountDown[3];
         CurrentCount.scale(CountScale);
      }
      else if (State.contains(dmz.consts.GameCountdown3)
            && !PreviousState.contains(dmz.consts.GameCountdown3)) {

         CDSwitch.enableSingleSwitchState(3);
         CurrentCount = CountDown[2];
         CurrentCount.scale(CountScale);
      }
      else if (State.contains(dmz.consts.GameCountdown2)
            && !PreviousState.contains(dmz.consts.GameCountdown2)) {

         CDSwitch.enableSingleSwitchState(2);
         CurrentCount = CountDown[1];
         CurrentCount.scale(CountScale);
      }
      else if (State.contains(dmz.consts.GameCountdown1)
            && !PreviousState.contains(dmz.consts.GameCountdown1)) {

         CDSwitch.enableSingleSwitchState(1);
         CurrentCount = CountDown[0];
         CurrentCount.scale(CountScale);
      }
      else if (State.contains(dmz.consts.GameActive)
            && !PreviousState.contains(dmz.consts.GameActive)) {

         CDSwitch.setSwitchStateAll(false);
         CurrentCount = null;
      }

      if (State.contains(dmz.consts.GameWaiting)
            && PreviousState.contains(dmz.consts.GameActive)) {

         GameOver.scale(0.01);
         GSwitch.enableSingleSwitchState(1, true);
         GoActive = true;
      }
   }
});
